src/hotspot/share/cds/dynamicArchive.cpp

```diff
@@ -349,6 +396,78 @@ class VM_PopulateDynamicDumpSharedSpace: public VM_GC_Sync_Operation {
   }
 };
 
+GrowableArray<ObjArrayKlass*>* DynamicArchive::_array_klasses = nullptr;
+Array<ObjArrayKlass*>* DynamicArchive::_dynamic_archive_array_klasses = nullptr;
+
+void DynamicArchive::append_array_klass(ObjArrayKlass* ak) {
+  if (_array_klasses == nullptr) {
+    _array_klasses = new (mtClassShared) GrowableArray<ObjArrayKlass*>(50, mtClassShared);
+  }
+  _array_klasses->append(ak);
+}
+
+void DynamicArchive::dump_array_klasses() {
+  assert(DynamicDumpSharedSpaces, "DynamicDumpSharedSpaces only");
+  if (_array_klasses != nullptr) {
+    ArchiveBuilder* builder = ArchiveBuilder::current();
+    int num_array_klasses = _array_klasses->length();
+    _dynamic_archive_array_klasses =
+        ArchiveBuilder::new_ro_array<ObjArrayKlass*>(num_array_klasses);
+    for (int i = 0; i < num_array_klasses; i++) {
+      ObjArrayKlass* oak = _array_klasses->at(i);
+      _dynamic_archive_array_klasses->at_put(i, oak);
```

Comment:
Isn't storing `oak` in the array  covered by the next call to `write_pointer_in_buffer`?

Reply:
Thanks for your review. The `write_pointer_in_buffer` writes the buffered address of `oak` to the array. So the `at_put` call isn't necessary.